Extend open_run to return a single run DataCollection of both raw and proc

philsmt opened the pull request:
The "single-run" logic introduced in #164 makes it somewhat awkward to obtain a `DataCollection` of both `raw` and `proc` for a single run.  Just using `open_run(..., data='raw').union(open_run(..., data='proc'))` with the same parameters no longer has the `is_single_run` flag. In an effort to no longer mindlessly copy over all raw data (and/or having the pre-calibrated data available, too!), this will be common use case.

While it can be set manually, this is confusing and unnecessarily verbose. This PR extends `open_run` with an option for `data` to create this union automatically and re-enable the `is_single_run` flag.

As an WIP, I've added the string `'all'` and the `Ellipsis`, but I don't have any strong opinions of this. Given this flag is only useful on our infrastructure and there won't be any new folder suddenly beyond the current ones, a string like `'all'` is probably safe?

tmichela commented on extra_data/reader.py:
```diff
@@ -1449,6 +1449,16 @@ def open_run(
+                      inc_suspect_trains=inc_suspect_trains)
+
+        dc = open_run(*args, data='raw', **kwargs).union(
+            open_run(*args, data='proc', **kwargs))
```
I'm afraid you're testing that in a very specific case. In practice there are keys duplication between raw and proc data, which will be conflicting when merging the 2 data collections.

philsmt replied:
Ouch, how ironic the very problem I'm trying to avoid with REMI is biting this solution now. I'll take a look whether one could add an `ignore_conflicts` flag to `DataCollection.union()`.

takluyver replied:
You could use `raw_extra = raw.deselect(proc)`

takluyver replied:
Just remembering, I think `.deselect()` works for this where it can drop entire files for the resulting DataCollection, as with the big detectors which are in their own files, but it won't work if it deselects some sources but still has other data from the same files, because `.union()` will effectively reselect these. I think this was part of why we postponed this before.

(I'm not sure I've explained that very clearly. I can have another go if it doesn't make sense)

philsmt replied:
Yes, I've just noticed this. When calibrated data adds a virtual source, `.deselect()` will throw an exception.
But since we filter on the source level here anyway, I might as well exclude those explicitly.

takluyver commented on extra_data/reader.py:
```diff
@@ -1449,6 +1449,16 @@ def open_run(
         files which don't have this flag, out-of-sequence train IDs are suspect.
         If True, it tries to include these trains.
     """
+    if data is Ellipsis or data == 'all':
```
I'd lean towards just one way (`data='all'`) to indicate this - `data=...` is cool but a bit cryptic.

(I think we used `data='proc'` rather than `corrected=True` because we anticipated one day adding an option to combine the two :+1: )

takluyver commented:
(Neat, Github automatically updated the target branch when I merged #164)

philsmt commented:
Okay, removed `Ellipsis` and added @takluyver 's beautiful solution to the duplicate sources problem. Works for me with both detector-type data (with _real_ calibrations) and my REMI data.

As a result from today's discussion, I will attempt to push for no longer copying raw data to `proc` and soft links in the interim. If this happens (and it will take time!), this will become really useful.

takluyver commented:
I think this should work where either the proc data contains entirely new sources (as we've agreed it should where we're designing it now), or where the sources which get corrected correspond to complete files (as for AGIPD, DSSC & LPD data, at least). It would still have issues if it needed to include part but not all of a raw file. I'm inclined to say that's good enough for the time being.

It would be good to add a quick test for this mechanism - it can be an extra piece inside `test_open_run` which already has everything it would need in place.

philsmt commented:
> It would still have issues if it needed to include part but not all of a raw file. I'm inclined to say that's good enough for the time being.

True, but I think this case does not exist in the wild. Traditional detectors have their own DA over xtdf or just a single source. As it happens, even in the REMI case, it's only the digitizer.
In the long run, the optimal solution might be a conflict resolution flag.

> It would be good to add a quick test for this mechanism - it can be an extra piece inside `test_open_run` which already has everything it would need in place.

Sure, coming in hopefully later today.

philsmt commented:
Ignore the failed CI please, I commited a test script by accident 🙈

takluyver commented on extra_data/tests/test_reader_mockdata.py:
```diff
@@ -701,6 +701,19 @@ def test_open_run(mock_spb_raw_run, mock_spb_proc_run, tmpdir):
+                else:
+                    # Non-AGIPD data is in raw.
+                    # (CAM, XGM)
+                    assert '/proc/' not in file.filename
```
Any particular reason to check `not in` in both cases? I forget now why I did one `in` and one `not in` in the code just above.

Is it worth checking that both kinds of sources are actually there? At the moment, this would pass if there were no sources at all, I think.

philsmt replied:
Ha, fair point with the sources check! But yes, I wanted to mirror your assertions.

philsmt replied:
Added a test of sources for `open_run(data='all')`. Strictly speaking, the mock SPB run is a special case where `raw` contains all sources and `proc` only a select few, but this should neatly cover the two real world examples of

a) copy of all `raw` data in `proc` for traditiona detectors and
b) additional sources in `proc` for REMI data

takluyver commented:
Thanks, this LGTM
